### ISSUE TITLE
Use Dask to run the cell simulations in parallel

### ISSUE THREAD

**Issue by wigging** [pybamm-team/liionpack#30]
This isn't really an issue but more of a feature request or enhancement. Would it be possible to use [Dask](https://dask.org) to run the cell simulations in parallel? I created a basic example (see below) of running the SPMe model for several discharges in parallel using Dask. Compare elapsed time with and without Dask by commenting out the appropriate section in `main()`. Elapsed times are given in the table below when running on an 8-core CPU.

Dask is made for massive parallelization and it's fairly easy to setup for CPU and GPU clusters. If it can be used with liionpack then it could provide a huge performance boost for large pack simulations. I haven't tried Casadi's parallel features beyond running on a single CPU but I don't think it will be easy to scale compared to using Dask.

| Example | Elapsed time |
| -------- | ------------- |
| No Dask |  8.57 seconds |
| Dask       |  3.83 seconds |

```python
import matplotlib.pyplot as plt
import pybamm
import time
from dask.distributed import Client

def generate_plots(discharge, t, capacity, current, voltage):

    def styleplot(ax):
        ax.legend(loc='best')
        ax.grid(color='0.9')
        ax.set_frame_on(False)
        ax.tick_params(color='0.9')

    _, ax = plt.subplots(tight_layout=True)
    for i in range(len(discharge)):
        ax.plot(t[i], current[i], label=f'{discharge[i]} A')
    ax.set_xlabel('Time [s]')
    ax.set_ylabel('Current [A]')
    styleplot(ax)

    _, ax = plt.subplots(tight_layout=True)
    for i in range(len(discharge)):
        ax.plot(t[i], voltage[i], label=f'{discharge[i]} A')
    ax.set_xlabel('Time [s]')
    ax.set_ylabel('Terminal voltage [V]')
    styleplot(ax)

    _, ax = plt.subplots(tight_layout=True)
    for i in range(len(discharge)):
        ax.plot(capacity[i], voltage[i], label=f'{discharge[i]} A')
    ax.set_xlabel('Discharge capacity [Ah]')
    ax.set_ylabel('Terminal voltage [V]')
    styleplot(ax)

    plt.show()

def run_simulation(dis, t_eval):

    model = pybamm.lithium_ion.SPMe()

    param = model.default_parameter_values
    param['Current function [A]'] = '[input]'

    sim = pybamm.Simulation(model, parameter_values=param)
    sim.solve(t_eval, inputs={'Current function [A]': dis})

    return sim.solution

def main(client):
    tic = time.perf_counter()

    discharge = [4, 3.5, 3, 2.5, 2, 1.8, 1.5, 1]  # discharge currents [A]
    t_eval = [0, 4000]                            # evaluation time [s]

    # No Dask
    # ------------------------------------------------------------------------

    # label = 'no Dask'

    # sols = []
    # for dis in discharge:
    #     sol = run_simulation(dis, t_eval)
    #     sols.append(sol)

    # Dask
    # ------------------------------------------------------------------------

    label = 'Dask'

    lazy_sols = client.map(run_simulation, discharge, t_eval=t_eval)
    sols = client.gather(lazy_sols)

    # ------------------------------------------------------------------------

    t = []
    capacity = []
    current = []
    voltage = []

    for sol in sols:
        t.append(sol['Time [s]'].entries)
        capacity.append(sol['Discharge capacity [A.h]'].entries)
        current.append(sol['Current [A]'].entries)
        voltage.append(sol["Terminal voltage [V]"].entries)

    toc = time.perf_counter()
    print(f'Elapsed time ({label}) = {toc - tic:.2f} s')

    generate_plots(discharge, t, capacity, current, voltage)

if __name__ == '__main__':
    client = Client()
    print(client)
    main(client)
```

**Comment by srikanthallu** [comment]
@TomTranter, could you elaborate some more on data dependencies between threads. Is it because we are doing a coupled solve with thermal ? Also, what was the library you used for shared memory parallel process ? Adaptive time stepping is important and will definitely help us accelerate.

**Comment by tinosulzer** [comment]
This would be a good first issue to tackle for understanding memory requirements https://github.com/pybamm-team/PyBaMM/issues/1442

**Comment by tinosulzer** [comment]
The way initial conditions are handled, especially when there are input parameters in the initial conditions, is also not perfect https://github.com/pybamm-team/PyBaMM/blob/6ccce9f817db8e5df69e6a3de90587b62e20180a/pybamm/solvers/base_solver.py#L1261-L1294

This might cause some difficulties with creating a single model and solving multiple times. That part of the code could do with a fresh pair of eyes tbh

**Comment by TomTranter** [comment]
Hey so I pushed some changes and I think basically you wanna try and put as much of the code in the for loop in `_serial_step` into a dask process

**Comment by wigging** [comment]
@TomTranter Dask is complaining that some of the variable types are `DM`. It looks like these are types returned from Casadi. Can you edit `_serial_step_dask` and `solve_dask` so there are no Casadi objects? I don't know how to support non-standard Python objects with Dask. I'm still learning how to use it.

**Comment by wigging** [comment]
Apparently objects must be able to be serialized for use with Dask. This may be causing the issues I'm having with passing Casadi and PyBaMM objects to the mapped Dask function. http://distributed.dask.org/en/latest/serialization.html

**Comment by TomTranter** [comment]
I think the integrator will work is you pass numpy arrays... but I'm not sure

**Comment by wigging** [comment]
@TomTranter I rebased the dask branch with main so everything is up-to-date with main. If you need to work on the dask branch make sure you pull (or force pull) the latest version to your local dask branch.

**Comment by TomTranter** [comment]
I'm getting issues again - maybe as I use sourcetree but I pulled down some changes in the dask branch and it's saying I have merge conflicts

I really really would prefer it if you didn't rebase

**Comment by TomTranter** [comment]
I just did a hard reset on my local machine to fix the issue
git reset --hard origin/dask

**Comment by TomTranter** [comment]
@wigging I made a little progress on this. I added a self contained script called dask debug with the bare minimal integrator functionality getting mapped. If you look at the dashboard there is a lot of waiting with nothing happening which I guess is data transfer in between steps. Maybe turning all the states into one big dask array and indexing into it would speed things up if dask handles these things efficiently in memory. If not we can try sharedarray

**Comment by wigging** [comment]
I'm getting close to a Dask solution. I ran the example using Python's `from concurrent.futures import ProcessPoolExecutor
` and chunked arrays which is giving good speed improvements. So now I'm adapting that approach for use with Dask. To be continued...

**Comment by TomTranter** [comment]
We now have a solution in the dask PR #66 but it's slower than just running the casadi integrator. I don't know how that mapped integrator works as the docs are very patchy and you can only get so far with debug but they must be doing something clever because it's pretty instantaneous to kick off. It looked like dask actors were going to be useful and might still be but are quite experimental. The idea is to persist an actor class on a cpu process which holds it's own state variables. This is basically an array of solutions for that actors chunk of batteries. It then steps through them individually and updates the solution but only passes back the output. This is more efficient for data transfer but it means keeping a lot of data in memory. So a new actor is created for each process and I guess there's some overhead with this. It doesn't seem to scale as well as casadi (about twice as slow to step solve excluding the extra time to set up the client and instantiate the actors). There is another package called ray which does something similar but their actor functionality is a bit more developed than dask and they also talk about ways to use shared memory between processes which may speed up sending inputs and outputs about. It's hard to profile the dask actors too as all their tasks are hidden from the task stream. I will check out Ray but perhaps going forward we should think about writing our own parallel wrapper to sundials... Maybe @martinjrobins has some thoughts on how long this would take?

**Comment by wigging** [comment]
Keep in mind that the reason to use Dask is to be able to scale to multi-node clusters (CPUs and GPUs). Can Casadi, Ray, etc. take advantage of such hardware or are they limited to a single machine?

**Comment by TomTranter** [comment]
Ray certainly can. We need to find out how the casadi map works. No answer from them on the forum yet

**Comment by TomTranter** [comment]
Addressed in pr #66
